### ISSUE TITLE
Flaky test TestAccGitlabProject_initializeWithReadme

### ISSUE THREAD

**Issue by armsnyder** [gitlabhq/terraform-provider-gitlab#800]
The `TestAccGitlabProject_initializeWithReadme` test has been failing on and off.

```
=== RUN   TestAccGitlabProject_initializeWithReadme
    resource_gitlab_project_test.go:328: Step 1/1 error: Check failed: Check 2/3 error: expected default branch to be protected
```

https://github.com/gitlabhq/terraform-provider-gitlab/runs/4928564294

**Comment by PatrickRice-KSC** [comment]
I'm starting to investigate this issue.

**Comment by PatrickRice-KSC** [comment]
Still working on reproducing locally. Current theory is a potential race condition between project creation and the initial commit behind the scenes. To try and suss out an error condition I've added some debug logging around project/commit times, and I've creating a perpetual loop to literally run the tests until it fails against a local gitlab instance using the following syntax: 
```
while make testacc RUN=TestAccGitlabProject_initializeWithReadme GITLAB_TOKEN=ACCTEST1234567890123 GITLAB_BASE_URL=http://127.0.0.1:8080/api/v4 TF_ACC=1 ; do sleep 1; done
```
I tend modified the makefile slightly to clear the test cache with each run, and I'm just going to let it run overnight to try and get a local failure with a bunch of debug logging.

**Comment by PatrickRice-KSC** [comment]
That didn't take a ton of time. It seems like GitLab is applying the branch protection at a given delay after project creation. I added a bit of debugging to the failure state to essentially wait 10s then re-read the branch protection, and it flips to true after the first wait. Likely we just need to add a slight read delay to the check and the test will be stable. Evidence:

```
go clean -testcache

TF_ACC=1 GITLAB_TOKEN=ACCTEST1234567890123 GITLAB_BASE_URL=http://127.0.0.1:8080/api/v4 go test -v ./gitlab -test.run TestAccGitlabProject_initializeWithReadme -timeout 40m

=== RUN   TestAccGitlabProject_initializeWithReadme

--- PASS: TestAccGitlabProject_initializeWithReadme (11.11s)

=== RUN   TestAccGitlabProject_initializeWithReadmeAndCustomDefaultBranch

--- PASS: TestAccGitlabProject_initializeWithReadmeAndCustomDefaultBranch (13.69s)

PASS

ok  	github.com/gitlabhq/terraform-provider-gitlab/gitlab	24.958s

go clean -testcache

TF_ACC=1 GITLAB_TOKEN=ACCTEST1234567890123 GITLAB_BASE_URL=http://127.0.0.1:8080/api/v4 go test -v ./gitlab -test.run TestAccGitlabProject_initializeWithReadme -timeout 40m

=== RUN   TestAccGitlabProject_initializeWithReadme

--- PASS: TestAccGitlabProject_initializeWithReadme (11.19s)

=== RUN   TestAccGitlabProject_initializeWithReadmeAndCustomDefaultBranch

--- PASS: TestAccGitlabProject_initializeWithReadmeAndCustomDefaultBranch (11.89s)

PASS

ok  	github.com/gitlabhq/terraform-provider-gitlab/gitlab	23.184s

go clean -testcache

TF_ACC=1 GITLAB_TOKEN=ACCTEST1234567890123 GITLAB_BASE_URL=http://127.0.0.1:8080/api/v4 go test -v ./gitlab -test.run TestAccGitlabProject_initializeWithReadme -timeout 40m

=== RUN   TestAccGitlabProject_initializeWithReadme

    resource_gitlab_project_test.go:973: [DEBUG] branch commit created time: 2022-02-07 05:05:14 +0000 UTC

    resource_gitlab_project_test.go:974: [DEBUG] branch protection: false

    resource_gitlab_project_test.go:975: [DEBUG] Default branch: true

    resource_gitlab_project_test.go:979: [DEBUG] branch protection: true

    resource_gitlab_project_test.go:979: [DEBUG] branch protection: true

    resource_gitlab_project_test.go:979: [DEBUG] branch protection: true

^CFAIL
```

Any issues with this approach @armsnyder or @timofurrer ?

**Comment by armsnyder** [comment]
@PatrickRice-KSC Thanks for your findings! I think the proposed fix of sleeping during the test would mask the bug but not address it. If the issue is GitLab needing to update behind the scenes after the API call, then the `resourceGitlabProjectCreate` function needs to be modified to wait for that to finish.

You could add a [WaitForState](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/helper/resource#StateChangeConf.WaitForState) call after protecting the branch, so that it waits for GitLab API to finish protecting the branch behind the scenes. There's an example of this in that same function where it handles a project import.

**Comment by PatrickRice-KSC** [comment]
Yep, makes sense. I'll get a PR open for that hers shortly. I'll likely need to add a counter to my loop script so I can let it run for 50-100 loops just so I feel good about it being resolved - it was taking upwards of 20-30 loops to encounter a fail state in my above testing, likely because the load on the instance determines how often the race condition happens and I was only hitting it with a couple tests.

**Comment by PatrickRice-KSC** [comment]
Just an update on this (I don't like going silent on items for too long) - I've got some code written locally here, but I'm still seeing a test failure when running the test repeatedly, so doing some more debugging before I open a PR to see if I just have the WaitForState in the wrong spot, or if there is a second issue I need to find.

**Comment by PatrickRice-KSC** [comment]
PR submitted. I initially thought the race condition may exist only within the nested block that modified the default branch, but in fact it exists within the project resource itself. Moving the WaitForState outside the default branch modification block caused the race condition to close - test passed in 300 / 300 iterations which is well beyond the percentage required to simulate a failure in the past.
